srai/embedders/hex2vec/embedder.py

```diff
@@ -0,0 +1,189 @@
+        """
+        Create region embeddings.
+
+        Args:
+            regions_gdf (gpd.GeoDataFrame): Region indexes and geometries.
+            features_gdf (gpd.GeoDataFrame): Feature indexes, geometries and feature values.
+            joint_gdf (gpd.GeoDataFrame): Joiner result with region-feature multi-index.
+
+        Returns:
+            pd.DataFrame: Embedding and geometry index for each region in regions_gdf.
+
+        Raises:
+            ValueError: If features_gdf is empty and self.expected_output_features is not set.
+            ValueError: If any of the gdfs index names is None.
+            ValueError: If joint_gdf.index is not of type pd.MultiIndex or doesn't have 2 levels.
+            ValueError: If index levels in gdfs don't overlap correctly.
+        """
+        import torch
+
+        self._check_is_fitted()
+        counts_df = self._get_raw_counts(regions_gdf, features_gdf, joint_gdf)
+        counts_tensor = torch.from_numpy(counts_df.values)
+        embeddings = self._model(counts_tensor).detach().numpy()  # type: ignore
+        return pd.DataFrame(embeddings, index=counts_df.index)
+
+    def fit(
+        self,
+        regions_gdf: gpd.GeoDataFrame,
+        features_gdf: gpd.GeoDataFrame,
+        joint_gdf: gpd.GeoDataFrame,
+        neighbourhood: Neighbourhood[T],
+        negative_sample_k_distance: int = 2,
+        batch_size: int = 32,
+        trainer_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        """
+        Fit the model to the data.
+
+        Args:
+            regions_gdf (gpd.GeoDataFrame): Region indexes and geometries.
+            features_gdf (gpd.GeoDataFrame): Feature indexes, geometries and feature values.
+            joint_gdf (gpd.GeoDataFrame): Joiner result with region-feature multi-index.
+            neighbourhood (Neighbourhood[T]): The neighbourhood to use.
+                Should be intialized with the same regions.
+            negative_sample_k_distance (int, optional): When sampling negative samples,
+                sample from a distance > k. Defaults to 2.
+            batch_size (int, optional): Batch size. Defaults to 32.
+            trainer_kwargs (Optional[Dict[str, Any]], optional): Trainer kwargs. Defaults to None.
+
+        Raises:
+            ValueError: If features_gdf is empty and self.expected_output_features is not set.
+            ValueError: If any of the gdfs index names is None.
+            ValueError: If joint_gdf.index is not of type pd.MultiIndex or doesn't have 2 levels.
+            ValueError: If index levels in gdfs don't overlap correctly.
+            ValueError: If negative_sample_k_distance < 2.
+        """
+        import pytorch_lightning as pl
+        from torch.utils.data import DataLoader
+
+        counts_df = self._get_raw_counts(regions_gdf, features_gdf, joint_gdf)
+        num_features = len(counts_df.columns)
+        self._model = Hex2VecModel(layer_sizes=[num_features, *self._encoder_sizes])
```

Comment:
We can't setup learning rate even as `**hex2vec_model_kwargs` in init, do we want to leave it like that?